File: src/main/java/ca/mcgill/science/tepid/server/util/SessionManager.java
package ca.mcgill.science.tepid.server.util;

import ca.mcgill.science.tepid.common.Session;
import ca.mcgill.science.tepid.common.User;
import ca.mcgill.science.tepid.common.Utils;
import ca.mcgill.science.tepid.common.ViewResultSet;
import ca.mcgill.science.tepid.server.rest.Sessions;
import in.waffl.q.Promise;
import in.waffl.q.Q;

import org.mindrot.jbcrypt.BCrypt;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.MediaType;

import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

public class SessionManager {

    private static SessionManager instance;
    private static final Logger logger = LoggerFactory.getLogger(SessionManager.class);

    public static synchronized SessionManager getInstance() {
        if (instance == null) instance = new SessionManager();
        return instance;
    }

    private static final WebTarget couchdb = CouchClient.getTepidWebTarget();

    private static class UserResultSet extends ViewResultSet<String, User> {
    }

    private SessionManager() {

    }

    public Session start(User user, int expiration) {
        Session s = new Session(Utils.newSessionId(), user, expiration);
        couchdb.path(s.getId()).request().put(Entity.entity(s, MediaType.APPLICATION_JSON));
        return s;
    }

    public
    @Nullable
    Session get(String id) {
        Session s = null;
        try {
            s = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(Session.class);
        } catch (Exception e) {
        }
        if (s != null && s.getExpiration().getTime() > System.currentTimeMillis()) {
            return s;
        } else {
            return null;
        }
    }

    /**
     * Check if session exists and isn't expired
     *
     * @param s sessionId
     * @return true for valid, false otherwise
     */
    public boolean valid(String s) {
        return this.get(s) != null;
    }

    public void end(String s) {
        Session over = couchdb.path(s).request(MediaType.APPLICATION_JSON).get(Session.class);
        couchdb.path(over.getId()).queryParam("rev", over.getRev()).request().delete(String.class);
        logger.debug("Ending session for {}.", over.getUser().longUser);
    }

    /**
     * Authenticate user is necessary
     *
     * @param sam short user
     * @param pw  password
     * @return authenticated user
     */
    public
    @Nullable
    User authenticate(String sam, String pw) {
        User dbUser = getSam(sam);
        if (dbUser != null && dbUser.authType != null && dbUser.authType.equals("local")) {
            if (BCrypt.checkpw(pw, dbUser.password)) {
                return dbUser;
            } else {
                return null;
            }
        } else {
            return Ldap.authenticate(sam, pw);
        }
    }

    /*
     * TODO compare method above against method below (from the old repo)
    public User authenticate(String sam, String pw) {
        User dbUser = null;
        try {
            if (sam.contains("@")) {
                UserResultSet results = couchdb.path("_design").path("main").path("_view").path("byLongUser").queryParam("key", "\""+sam.replace("@", "%40")+"\"").request(MediaType.APPLICATION_JSON).get(UserResultSet.class);
                if (!results.rows.isEmpty()) dbUser = results.rows.get(0).value;
            } else {
                dbUser = couchdb.path("u" + sam).request(MediaType.APPLICATION_JSON).get(User.class);
            }
        } catch (Exception e) {}
        if (dbUser != null && dbUser.authType != null && dbUser.authType.equals("local")) {
            if (BCrypt.checkpw(pw, dbUser.password)) {
                return dbUser;
            } else {
                return null;
            }
        } else {
            return Ldap.authenticate(sam, pw);
        }
    }
    */

    /**
     * Retrieve user from Ldap if available, otherwise retrieves from cache
     *
     * @param sam short user
     * @param pw  password
     * @return user if found
     * @see #queryUserCache(String)
     */
    public
    @Nullable
    User queryUser(String sam, String pw) {
        if (Ldap.LDAP_ENABLED) return Ldap.queryUser(sam, pw);
        return queryUserCache(sam);
    }

    private
    @Nullable
    User getSam(String sam) {
        try {
            if (sam.contains("@")) {
                UserResultSet results = couchdb.path("_design").path("main").path("_view").path("byLongUser").queryParam("key", "\"" + sam.replace("@", "%40") + "\"").request(MediaType.APPLICATION_JSON).get(UserResultSet.class);
                if (!results.rows.isEmpty()) return results.rows.get(0).value;
            } else {
                return couchdb.path("u" + sam).request(MediaType.APPLICATION_JSON).get(User.class);
            }
        } catch (Exception ignored) {
        }
        return null;
    }

    /**
     * Get user if exists and sets salutation
     *
     * @param sam shortId
     * @return user if exists
     */
    public
    @Nullable
    User queryUserCache(String sam) {
        User dbUser = getSam(sam);
        if (dbUser == null) return null;
        dbUser.salutation = dbUser.nick == null ?
                ((dbUser.preferredName != null && !dbUser.preferredName.isEmpty())
                        ? dbUser.preferredName.get(dbUser.preferredName.size() - 1) : dbUser.givenName) : dbUser.nick;
        return dbUser;
    }

    /**
     * Sends list of matching {@link User}s based on current query
     *
     * @param like  prefix
     * @param limit max list size
     * @return list of matching users
     */
    public Promise<List<User>> autoSuggest(String like, int limit) {
        if (!Ldap.LDAP_ENABLED) {
            Q<List<User>> emptyPromise = Q.defer();
            emptyPromise.resolve(Arrays.asList(new User[0]));
            return emptyPromise.promise;
        }
        return Ldap.autoSuggest(like, limit);
    }

    /**
     * Retrieves user role
     *
     * @param u user to check
     * @return String for role
     */
    public String getRole(User u) {
        if (u == null) return null;
        Calendar cal = Calendar.getInstance();
        String[] elderGroups = {"***REMOVED***"},
                userGroups = {"***REMOVED***", "***REMOVED***" + cal.get(Calendar.YEAR) + (cal.get(Calendar.MONTH) < 8 ? "W" : "F")},
                ctferGroups = {"***REMOVED***", "***REMOVED***"};
        if (u.authType == null || !u.authType.equals("local")) {
            if (u.groups == null) return null;
            for (String g : elderGroups) {
                if (u.groups.contains(g)) return "elder";
            }
            for (String g : ctferGroups) {
                if (u.groups.contains(g)) return "ctfer";
            }
            for (String g : userGroups) {
                if (u.groups.contains(g)) return "user";
            }
            return null;
        } else {
            if (u.role != null && u.role.equals("admin")) {
                return "elder";
            } else {
                return "user";
            }
        }
    }

    /**
     * Sets exchange student status
     *
     * @param sam      shortId
     * @param exchange boolean for exchange status
     */
    public void setExchangeStudent(String sam, boolean exchange) {
        if (Ldap.LDAP_ENABLED) Ldap.setExchangeStudent(sam, exchange);
    }

}

File: src/main/java/ca/mcgill/science/tepid/server/util/SessionManager.kt
package ca.mcgill.science.tepid.server.util

import `in`.waffl.q.Promise
import `in`.waffl.q.Q
import ca.mcgill.science.tepid.models.bindings.LOCAL
import ca.mcgill.science.tepid.models.bindings.withDbData
import ca.mcgill.science.tepid.models.data.FullSession
import ca.mcgill.science.tepid.models.data.FullUser
import ca.mcgill.science.tepid.models.data.User
import ca.mcgill.science.tepid.utils.WithLogging
import com.fasterxml.jackson.databind.node.ObjectNode
import org.mindrot.jbcrypt.BCrypt
import java.math.BigInteger
import java.security.SecureRandom

/**
 * SessionManager is responsible for managing sessions and dealing with the underlying authentication.
 * It is analogous to PAM, in that everything which needs authentication or user querying goes through this.
 * For managing sessions, it can start, resume, and end sessions
 * For user querying, it first checks the DB cache. The cache is updated every time a query to the underlying authentication is made.
 * Since it also provides an interface with the underlying authentication, it also provides username autosuggestion and can set users as exchange.
 */

object SessionManager : WithLogging() {

    private const val HOUR_IN_MILLIS = 60 * 60 * 1000
    private val numRegex = Regex("[0-9]+")
    private val shortUserRegex = Regex("[a-zA-Z]+[0-9]*")

    private val random = SecureRandom()

    fun start(user: FullUser, expiration: Int): FullSession {
        val session = FullSession(user = user, expiration = System.currentTimeMillis() + expiration * HOUR_IN_MILLIS)
        val id = BigInteger(130, random).toString(32)
        session._id = id
        log.trace("Creating session $id")
        val out = CouchDb.path(id).putJson(session)
        println(out)
        return session
    }

    operator fun get(token: String): FullSession? {
        val session = CouchDb.path(token).getJsonOrNull<FullSession>() ?: return null
        if (session.isValid()) return session
        log.trace("Session $token is invalid; now ${System.currentTimeMillis()} expiration ${session.expiration}; deleting")
        CouchDb.path(token).deleteRev()
        return null
    }

    /**
     * Check if session exists and isn't expired
     *
     * @param s sessionId
     * @return true for valid, false otherwise
     */
    fun valid(s: String): Boolean = this[s] != null

    fun end(s: String) {
        //todo test
        CouchDb.path(s).deleteRev()
    }

    /**
     * Authenticate user if necessary
     *
     * @param sam short user
     * @param pw  password
     * @return authenticated user
     */
    fun authenticate(sam: String, pw: String): FullUser? {
        val dbUser = queryUserDb(sam)
        log.trace("Db data for $sam")
        if (dbUser?.authType == LOCAL) {
            return if (BCrypt.checkpw(pw, dbUser.password)) dbUser else null
        }
        else if (Config.LDAP_ENABLED) {
            var ldapUser =  Ldap.authenticate(sam, pw)
            if (ldapUser!=null){
                ldapUser = mergeUsers(ldapUser, dbUser)
                updateDbWithUser(ldapUser)
            }
            return ldapUser
        }
        else {
            return null
        }
    }

    /**
     * Retrieve user from DB if available, otherwise retrieves from LDAP
     *
     * @param sam short user
     * @param pw  password
     * @return user if found
     */
    fun queryUser(sam: String?, pw: String?): FullUser? {
        if (sam == null) return null
        log.trace("Querying user $sam")

        val dbUser = queryUserDb(sam)

        if (dbUser != null) return dbUser

        if (Config.LDAP_ENABLED) {
            if (!sam.matches(shortUserRegex)) return null // cannot query without short user
            var ldapUser = Ldap.queryUserLdap(sam, pw) ?: return null

            ldapUser = mergeUsers(ldapUser, dbUser)

            if (dbUser != ldapUser) {
                updateDbWithUser(ldapUser)
            } else {
                log.trace("Not updating dbUser; already matches ldap user")
            }

            log.trace("Found user from ldap $sam: ${ldapUser.longUser}")
            return ldapUser
        }
        //finally
        return null
    }

    /**
     * Update [ldapUser] with db data
     * [queryAsOwner] should be true if [ldapUser] was retrieved by the owner rather than a resource account
     */
    internal fun mergeUsers(ldapUser: FullUser, dbUser: FullUser?): FullUser {
        // ensure that short users actually match before attempting any merge
        val ldapShortUser = ldapUser.shortUser ?: throw RuntimeException ("LDAP user does not have a short user. Maybe this will help {\"ldapUser\":\"${ldapUser.toString()},\"dbUser\":\"${dbUser.toString()}\"}")
        if (dbUser == null) return ldapUser
        if (ldapShortUser != dbUser.shortUser) throw RuntimeException ("Attempt to merge to different users {\"ldapUser\":\"${ldapUser.toString()},\"dbUser\":\"${dbUser.toString()}\"}")
        // proceed with data merge
        val newUser = ldapUser.copy()
        newUser.withDbData(dbUser)
        newUser.studentId = if (ldapUser.studentId!=-1) ldapUser.studentId else dbUser.studentId
        newUser.preferredName = dbUser.preferredName
        newUser.nick = dbUser.nick
        newUser.colorPrinting = dbUser.colorPrinting
        newUser.jobExpiration = dbUser.jobExpiration
        return newUser
    }
    /**
     * Uploads a [user] to the DB,
     * with logging for failures
     */
    internal fun updateDbWithUser(user: FullUser) {
        log.trace("Update db instance")
        try {
            val response = CouchDb.path("u${user.shortUser}").putJson(user)
            if (response.isSuccessful) {
                val responseObj = response.readEntity(ObjectNode::class.java)
                val newRev = responseObj.get("_rev")?.asText()
                if (newRev != null && newRev.length > 3) {
                    user._rev = newRev
                    log.trace("New rev for ${user.shortUser}: $newRev")
                }
            } else {
                log.error("Response failed: $response")
            }
        } catch (e1: Exception) {
            log.error("Could not put ${user.shortUser} into db", e1)
        }
    }


    /**
     * Retrieve a [FullUser] directly from the database when supplied with either a
     * short user, long user, or student id
     */
    fun queryUserDb(sam: String?): FullUser? {
        sam ?: return null
        val dbUser = when {
            sam.contains(".") ->
                CouchDb
                    .path(CouchDb.CouchDbView.ByLongUser)
                    .queryParam("key", "\"${sam.substringBefore("@")}%40${Config.ACCOUNT_DOMAIN}\"")
                    .getViewRows<FullUser>()
                    .firstOrNull()
            sam.matches(numRegex) ->
                CouchDb
                    .path(CouchDb.CouchDbView.ByStudentId)
                    .queryParam("key", sam)
                    .getViewRows<FullUser>()
                    .firstOrNull()
            else -> CouchDb.path("u$sam").getJsonOrNull()
        }
        dbUser?._id ?: return null
        log.trace("Found db user (${dbUser._id}) ${dbUser.displayName} for $sam")
        return dbUser
    }

    /**
     * Sends list of matching [User]s based on current query
     *
     * @param like  prefix
     * @param limit max list size
     * @return list of matching users
     */
    fun autoSuggest(like: String, limit: Int): Promise<List<FullUser>> {
        if (!Config.LDAP_ENABLED) {
            val emptyPromise = Q.defer<List<FullUser>>()
            emptyPromise.resolve(emptyList())
            return emptyPromise.promise
        }
        return Ldap.autoSuggest(like, limit)
    }

    /**
     * Sets exchange student status
     *
     * @param sam      shortId
     * @param exchange boolean for exchange status
     */
    fun setExchangeStudent(sam: String, exchange: Boolean) {
        if (Config.LDAP_ENABLED) Ldap.setExchangeStudent(sam, exchange)
    }

}
